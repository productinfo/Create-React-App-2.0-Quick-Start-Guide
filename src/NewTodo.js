import React, { Component } from "react";
import "./NewTodo.css";

class NewTodo extends Component {
  constructor(props) {
    super(props);
    this.state = { item: "" };
    this.handleUpdate = this.handleUpdate.bind(this);
    this.addTodo = this.addTodo.bind(this);
  }
  handleUpdate(event) {
    this.setState({ item: event.target.value });
  }
  addTodo() {
    this.props.addTodo(this.state.item);
    this.setState({ item: "" });
  }
  render() {
    return (
      <div className="NewTodo">
        <input
          type="text"
          onChange={this.handleUpdate}
          value={this.state.item}
        />
        &nbsp;&nbsp;
        <button onClick={this.addTodo}>Add</button>
      </div>
    );
  }
}

export default NewTodo;
