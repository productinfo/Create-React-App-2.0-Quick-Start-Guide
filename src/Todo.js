import React, { Component } from "react";
import "./Todo.css";

class Todo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.description,
      done: false
    };
    this.markAsDone = this.markAsDone.bind(this);
    this.removeTodo = this.removeTodo.bind(this);
  }
  markAsDone() {
    this.setState({ done: true });
  }
  cssClasses() {
    let classes = ["Todo"];
    if (this.state.done) {
      classes = [...classes, "Done"];
    }
    return classes.join(" ");
  }
  removeTodo() {
    this.props.removeTodo(this.state.description);
  }
  render() {
    return (
      <div className={this.cssClasses()}>
        {this.state.description}
        <br />
        <button className="MarkDone" onClick={this.markAsDone}>
          Mark as Done
        </button>
        <button className="RemoveTodo" onClick={this.removeTodo}>
          Remove Me
        </button>
      </div>
    );
  }
}

export default Todo;
