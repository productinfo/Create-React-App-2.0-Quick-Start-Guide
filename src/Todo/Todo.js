import React, { Component } from "react";
import { Button, ButtonGroup } from "reactstrap";

import styles from "./Todo.module.scss";

class Todo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.description,
      done: false,
      critical: false
    };

    this.markAsDone = this.markAsDone.bind(this);
    this.removeTodo = this.removeTodo.bind(this);
    this.markCritical = this.markCritical.bind(this);
  }
  markAsDone() {
    this.setState({ done: true });
  }
  markCritical() {
    this.setState({ critical: true });
  }
  cssClasses() {
    let classes = [];
    if (this.state.critical) {
      classes = [styles.critical];
    } else {
      classes = [styles.todo];
    }
    if (this.state.done) {
      classes = [...classes, styles.done];
    }
    return classes.join(" ");
  }
  removeTodo() {
    this.props.removeTodo(this.state.description);
  }
  render() {
    return (
      <div className={this.cssClasses()}>
        {this.state.description}
        <br />
        <hr className={styles.hr} />
        <ButtonGroup>
          <Button
            className="MarkDone"
            onClick={this.markAsDone}
            color="success"
          >
            Mark as Done
          </Button>
          <Button
            className="RemoveTodo"
            onClick={this.removeTodo}
            color="warning"
          >
            Remove Me
          </Button>
          <Button
            className="MarkCritical"
            onClick={this.markCritical}
            color="danger"
          >
            Mark as Critical
          </Button>
        </ButtonGroup>
      </div>
    );
  }
}

export default Todo;
