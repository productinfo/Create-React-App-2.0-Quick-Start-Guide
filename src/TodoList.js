import React, { Fragment, Component } from "react";
import Todo from "./Todo";
import "./TodoList.css";

import NewTodo from "./NewTodo";
import Divider from "./Divider";

class TodoList extends Component {
  constructor(props) {
    super(props);
    const [item1, item2, ...rest] = [
      "Write some code",
      "Change the world",
      "Take a nap",
      "Eat a cookie"
    ];
    this.state = {
      items: [item1, item2, rest.join(" and ")]
    };

    this.addTodo = this.addTodo.bind(this);
    this.removeTodo = this.removeTodo.bind(this);
  }
  addTodo(item) {
    this.setState({ items: [...this.state.items, item] });
  }
  removeTodo(removeItem) {
    const filteredItems = this.state.items.filter(description => {
      return description !== removeItem;
    });
    this.setState({ items: filteredItems });
  }
  renderItems() {
    return this.state.items.map(description => (
      <Fragment key={"item-" + description}>
        <Todo
          key={description}
          description={description}
          removeTodo={this.removeTodo}
        />
        <Divider key={"divide-" + description} />
      </Fragment>
    ));
  }
  render() {
    return (
      <div className="TodoList">
        <NewTodo addTodo={this.addTodo} />
        {this.renderItems()}
      </div>
    );
  }
}

export default TodoList;
