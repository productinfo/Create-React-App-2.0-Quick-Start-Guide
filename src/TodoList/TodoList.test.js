import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import TodoList from "./TodoList";
import NewTodo from "../NewTodo/NewTodo";
import Todo from "../Todo/Todo";

describe(TodoList, () => {
  const component = shallow(<TodoList />);

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<TodoList />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders and matches our snapshot", () => {
    const component = renderer.create(<TodoList />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a TodoList component", () => {
    expect(component.contains(<div className="TodoList" />));
  });

  it("includes a NewTodo component", () => {
    expect(component.find(NewTodo)).toHaveLength(1);
  });

  it("renders the correct number of Todo components", () => {
    const todoCount = component.state("items").length;
    expect(component.find(Todo)).toHaveLength(todoCount);
  });

  it("adds another Todo when the addTodo function is called", () => {
    const before = component.find(Todo).length;
    component.instance().addTodo("A new item");
    const after = component.find(Todo).length;
    expect(after).toBeGreaterThan(before);
  });

  it("removes a Todo from the list when the remove todo function is called", () => {
    const before = component.find(Todo).length;
    const removeMe = component.state("items")[0];
    component.instance().removeTodo(removeMe);
    const after = component.find(Todo).length;
    expect(after).toBeLessThan(before);
  });
});
