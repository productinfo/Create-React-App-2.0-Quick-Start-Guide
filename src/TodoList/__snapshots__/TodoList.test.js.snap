// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList renders and matches our snapshot 1`] = `
<div
  className="TodoList"
>
  <div
    className="NewTodo"
  >
    <div
      className="input-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Input item name here..."
        type="text"
        value=""
      />
      <button
        aria-label={null}
        className="btn btn-primary"
        onClick={[Function]}
        type="button"
      >
        Add
      </button>
    </div>
  </div>
  <div
    className="todo"
  >
    Write some code
    <br />
    <hr
      className="hr"
    />
    <div
      className="btn-group"
      role="group"
    >
      <button
        aria-label={null}
        className="MarkDone btn btn-success"
        onClick={[Function]}
        type="button"
      >
        Mark as Done
      </button>
      <button
        aria-label={null}
        className="RemoveTodo btn btn-warning"
        onClick={[Function]}
        type="button"
      >
        Remove Me
      </button>
      <button
        aria-label={null}
        className="MarkCritical btn btn-danger"
        onClick={[Function]}
        type="button"
      >
        Mark as Critical
      </button>
    </div>
  </div>
  <div
    className="Divider"
  >
    <hr />
  </div>
  <div
    className="todo"
  >
    Change the world
    <br />
    <hr
      className="hr"
    />
    <div
      className="btn-group"
      role="group"
    >
      <button
        aria-label={null}
        className="MarkDone btn btn-success"
        onClick={[Function]}
        type="button"
      >
        Mark as Done
      </button>
      <button
        aria-label={null}
        className="RemoveTodo btn btn-warning"
        onClick={[Function]}
        type="button"
      >
        Remove Me
      </button>
      <button
        aria-label={null}
        className="MarkCritical btn btn-danger"
        onClick={[Function]}
        type="button"
      >
        Mark as Critical
      </button>
    </div>
  </div>
  <div
    className="Divider"
  >
    <hr />
  </div>
  <div
    className="todo"
  >
    Take a nap and Eat a cookie
    <br />
    <hr
      className="hr"
    />
    <div
      className="btn-group"
      role="group"
    >
      <button
        aria-label={null}
        className="MarkDone btn btn-success"
        onClick={[Function]}
        type="button"
      >
        Mark as Done
      </button>
      <button
        aria-label={null}
        className="RemoveTodo btn btn-warning"
        onClick={[Function]}
        type="button"
      >
        Remove Me
      </button>
      <button
        aria-label={null}
        className="MarkCritical btn btn-danger"
        onClick={[Function]}
        type="button"
      >
        Mark as Critical
      </button>
    </div>
  </div>
  <div
    className="Divider"
  >
    <hr />
  </div>
</div>
`;
